src/hip/hip_build_utils.cpp

```diff
@@ -118,8 +126,60 @@ static boost::filesystem::path HipBuildImpl(boost::optional<TmpDir>& tmp_dir,
             WriteFile(inc_src, inc_path / inc_file);
         }
     }
-    src += "\nint main() {}\n";
-    WriteFile(src, tmp_dir->path / filename);
+
+    // Invoke mlir kernel generator if filename has mlir in it
+    if(filename.find("mlir_gen_igemm_conv2d_cpp") != std::string::npos)
```

Comment:
This code transforms `params.c_str()` to the HIP source code. So input is not HIP. Definitely not a job for "HipBuild". This should be moved to a separate function, most likely in separate module.

Most likely, we shall introduce another kind of kernel source, "MLIR generator input (?)" and implement accordingly.

Please expect more details in next review. Thanks.

Reply:
I can definitely come up with a new function to extract this section of if condition away. 

> Most likely, we shall introduce another kind of kernel source, "MLIR generator input (?)" and implement accordingly.

For now I'd suggest to not over-optimize. According to jack, we are not investing more effort in the cpp backend. All future kernels will be populated from the bin backend. I think it makes better sense we make the bin backend build path more robust.

Reply:
This is not an optimization. This is an effort to keep the design of the library clear and uniform. "MLIR binary backend" will be built re-using the same concepts as we will use in "cpp" backend. Please expect more info later, I still need time for reviewing.

Reply:
@atamazov Sounds good. Since reviewing this PR I believe you got a good sense of what I'm about to implement as well in the bin backend as well. It will be a combination of three new solvers + plumbing code. Since bin backend is capable of populating the binary altogether, it actually make the plumbing easier (with .so ended as kernel name). Honestly speaking, I wouldn't mind back out the changes related with only cpp backend, in the near future.

Reply:
This is not resolved.

Reply:
Hmm, I thought we are talking about writing the igemm logic in a seperate subroutine, which result in this GenIgemmHost() function: https://github.com/ROCmSoftwarePlatform/MIOpen/pull/676/commits/c209376880fa7ed0c94947936ab83e043b588c82#diff-9da6f2a9c82c3124e52b00d65a9c4a4d40fc66ec5b449e30719ebb9c07daaf78R104-R149

Does this not resolve your concern in here?

Reply:
Resolved.